cpp/include/Ice/DefaultSliceLoader.h

```diff
@@ -0,0 +1,165 @@
+        /// @return The single instance of DefaultSliceLoader.
+        static DefaultSliceLoaderPtr instance();
+
+        virtual ~DefaultSliceLoader();
+
+        Ice::ValuePtr newClassInstance(std::string_view typeId) const final;
+        std::exception_ptr newExceptionInstance(std::string_view typeId) const final;
+
+        template<class T> void addClass(int compactId)
+        {
+            std::lock_guard lock{_mutex};
+            auto p = _classFactories.find(T::ice_staticId());
+            if (p == _classFactories.end())
+            {
+                _classFactories[T::ice_staticId()] = {[] { return std::make_shared<T>(); }, 1};
+            }
+            else
+            {
+                p->second.second++;
+            }
+
+            if (compactId >= 0)
+            {
+                _compactIdTable[compactId] = T::ice_staticId();
+
+                std::string compactIdStr{std::to_string(compactId)};
+                p = _classFactories.find(compactIdStr);
+                if (p == _classFactories.end())
+                {
+                    _classFactories[compactIdStr] = {[] { return std::make_shared<T>(); }, 1};
+                }
+                else
+                {
+                    p->second.second++;
+                }
+            }
+        }
+
+        void removeClass(std::string_view typeId, int compactId)
+        {
+            std::lock_guard lock{_mutex};
+            auto p = _classFactories.find(typeId);
+            if (p != _classFactories.end())
+            {
+                if (--p->second.second == 0)
+                {
+                    _classFactories.erase(p);
+                }
+            }
+
+            if (compactId >= 0)
+            {
+                std::string compactIdStr{std::to_string(compactId)};
+                p = _classFactories.find(compactIdStr);
+                if (p != _classFactories.end())
+                {
+                    if (--p->second.second == 0)
+                    {
+                        _classFactories.erase(p);
+                    }
+                }
+            }
+        }
+
+        template<class T> void addException()
+        {
+            std::lock_guard lock{_mutex};
+            auto p = _exceptionFactories.find(T::ice_staticId());
+            if (p == _exceptionFactories.end())
+            {
+                _exceptionFactories[T::ice_staticId()] = {[] { return std::make_exception_ptr(T{}); }, 1};
+            }
+            else
+            {
+                p->second.second++;
+            }
+        }
+
+        void removeException(std::string_view typeId)
+        {
+            std::lock_guard lock{_mutex};
+            auto p = _exceptionFactories.find(typeId);
+            if (p != _exceptionFactories.end())
+            {
+                if (--p->second.second == 0)
+                {
+                    _exceptionFactories.erase(p);
+                }
+            }
+        }
+
+        // temporary
```

Comment:
This is needed until we can eliminate the compact ID resolver used by Python etc.